iceoryx_examples/automotive_soa/src/runtime.cpp

```diff
@@ -0,0 +1,176 @@
+    if (verifyThatServiceIsComplete(iceoryxServiceContainer))
+    {
+        autosarServiceContainer.push_back({serviceIdentifier, instanceIdentifier});
+    }
+
+    return autosarServiceContainer;
+    //! [verify iceoryx mapping]
+}
+
+kom::FindServiceCallbackHandle
+Runtime::EnableFindServiceCallback(kom::FindServiceCallback<kom::ProxyHandleType> handler,
+                                   kom::ServiceIdentifier& serviceIdentifier,
+                                   kom::InstanceIdentifier& instanceIdentifier) noexcept
+{
+    // Duplicate entries for the same service are allowed
+    if (!m_callbacks.push_back(CallbackEntryType(
+            handler, {serviceIdentifier, instanceIdentifier}, NumberOfAvailableServicesOnLastSearch())))
+    {
+        std::cerr << "Callback vector capacity exceeded, did not EnableFindServiceCallback!" << std::endl;
+        std::exit(EXIT_FAILURE);
+    }
+
+    //! [attach discovery to listener]
+    if (m_callbacks.size() == 1)
+    {
+        auto invoker = iox::popo::createNotificationCallback(invokeCallback, *this);
+        m_listener.attachEvent(m_discovery, iox::runtime::ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED, invoker)
+            .expect("Unable to attach discovery!");
+    }
+    //! [attach discovery to listener]
+
+    return kom::FindServiceCallbackHandle({serviceIdentifier, instanceIdentifier});
+}
+
+void Runtime::DisableFindServiceCallback(kom::FindServiceCallbackHandle handle) noexcept
+{
+    auto iter = m_callbacks.begin();
+    for (; iter != m_callbacks.end(); iter++)
+    {
+        if (std::get<1>(*iter).GetServiceId() == handle.GetServiceId()
+            && std::get<1>(*iter).GetInstanceId() == handle.GetInstanceId())
+        {
+            m_callbacks.erase(iter);
+            break;
+        }
+    }
+
+    if (m_callbacks.empty())
+    {
+        m_listener.detachEvent(m_discovery, iox::runtime::ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED);
+    }
+}
+
+bool Runtime::verifyThatServiceIsComplete(kom::ServiceHandleContainer<kom::ProxyHandleType>& container) noexcept
+{
+    // The service level of AUTOSAR Adaptive is not available in iceoryx, instead every publisher and server is
+    // considered as a service. A ara::com binding implementer would typically query the AUTOSAR meta model here, to
+    // find out if all event, fields and methods of a service are available. For the example we assume that the
+    // 'MinimalSkeleton' service is complete when the container contains the four iceoryx services:
+    //
+    // 1.    EventPublisher: MinimalSkeleton, Instance, Event  (iox::popo::Publisher)
+    // 2. a) FieldPublisher: MinimalSkeleton, Instance, Field  (iox::popo::Publisher)
+    // 2. b) FieldPublisher: MinimalSkeleton, Instance, Field  (iox::popo::Server)
+    // 3.    MethodServer:   MinimalSkeleton, Instance, Method (iox::popo::Server)
+
+    if (container.size() == 4U)
+    {
+        return true;
+    }
+    return false;
```

Comment:
Instead of writing 5 lines you could write one: `return (container.size() == 4U);`. This is just unnecessary complexity.